### ISSUE TITLE
CMake build directories break every update

### ISSUE THREAD

**Issue by nadiaholmquist** [Homebrew/homebrew-core#84312]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
% brew config
HOMEBREW_VERSION: 3.2.10
ORIGIN: https://github.com/Homebrew/brew
HEAD: ab2118a402a44451b1d7559167b6956dc47cc19a
Last commit: 22 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: bf1142ae8f34ecf5c9d372c13d0b9635ce35fc51
Core tap last commit: 18 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: quad-core 64-bit skylake
Clang: 13.0.0 build 1300
Git: 2.30.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.77.0 => /usr/bin/curl
macOS: 12.0-x86_64
CLT: 13.0.0.0.1.1628499445
Xcode: 13.0 => /Applications/Xcode-beta.app/Contents/Developer

% brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: "config" scripts exist outside your system or Homebrew directories.
`./configure` scripts often look for *-config scripts to determine if
software packages are installed, and which additional flags to use when
compiling and linking.

Having additional scripts in your path can confuse software installed via
Homebrew if the config script overrides a system or Homebrew-provided
script of the same name. We found the following "config" scripts:
  /opt/devkitpro/tools/bin/pkg-config
  /opt/devkitpro/tools/bin/x86_64-apple-darwin15-pkg-config

Warning: You are using macOS 12.
We do not provide support for this pre-release version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
pre-release version.


Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc
```

### 

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

Incrementally build a CMake project after having updated CMake from Homebrew.

### What happened (include all command output)?

```
nhp@narshe NanoboyAdvance % cmake --build build
gmake: /usr/local/Cellar/cmake/3.21.1/bin/cmake: No such file or directory
gmake: *** [Makefile:243: cmake_check_build_system] Error 127
```

### What did you expect to happen?

It picks up the new version of CMake automatically/does not rely on the specific versioned CMake directory from /usr/local/Cellar.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
1. `brew install cmake`
2. Clone a project that uses CMake, and create a build directory like `cmake -B build`
3. `brew upgrade` when there's a new version of CMake, or install another version some other way
4. CMake-generated Makefiles now can't find CMake
```

**Comment by SMillerDev** [comment]
I'd consider this a bug in CMake if it hardcodes the path to a command into the build directory.

**Comment by alebcay** [comment]
Consider reading https://stackoverflow.com/questions/61865483/brew-upgrade-cmake-but-the-old-version-is-still-called for further info.

My understanding is that the Makefile generated by CMake is ephemeral, it's not meant to be kept around and used incrementally later. I've never encountered a CMake-based project that shipped with a generated Makefile; users are expected to have CMake installed and generate the Makefile themselves when they want to build it (unlike Autoconf, where developers may run `autoconf`/`autoreconf` once and just ship the generated `configure`/`Makefile`; in Autoconf users rarely have to have Autoconf/Automake installed to perform the build).

If there are dependencies in the project that link to Homebrew libraries, they'll probably be broken too in the Makefile due to how Homebrew does versioning. To change this behavior/design likely requires re-designing the entire Homebrew Cellar system or also probably some non-trivial changes on CMake's end.

The workaround is simple though - just delete `CMakeCache.txt` and run the CMake configure step again.

**Comment by nadiaholmquist** [comment]
I'm not trying to ship a CMake-generated Makefile, I know you're not supposed to do that. The problem was for build directories of projects on the same machine, where it gets annoying having to type something like `cmake -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH="$(brew --prefix qt@6);$(brew --prefix libarchive)" -DUSE_QT6=ON -DMACOS_BUNDLE_LIBS=ON` every time a CMake update happens, which seems to lately be quite often.

**Comment by MikeMcQuaid** [comment]
We should probably figure out how to make CMake reference `/usr/local/opt/cmake/bin/cmake` rather than `/usr/local/Cellar/cmake/3.21.1/bin/cmake` in `CMakeCache.txt`.

**Comment by alebcay** [comment]
Looks like the relevant issue has been opened on the CMake side at https://gitlab.kitware.com/cmake/cmake/-/issues/21059.

From this issue, it sounds like there's no way to set this at build time right now. Someone opened https://gitlab.kitware.com/cmake/cmake/-/merge_requests/3929 as a proposed fix but it was not accepted.

**Comment by MikeMcQuaid** [comment]
> From this issue, it sounds like there's no way to set this at build time right now. Someone opened https://gitlab.kitware.com/cmake/cmake/-/merge_requests/3929 as a proposed fix but it was not accepted.

It looks like it just needs someone to follow up with the requested changes.

**Comment by carlocab** [comment]
As a workaround, try setting `CMAKE_COMMAND` to `$(brew --prefix cmake)/bin/cmake` when you invoke CMake. For example:
```
cmake -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH="$(brew --prefix qt@6);$(brew --prefix libarchive)" -DUSE_QT6=ON -DMACOS_BUNDLE_LIBS=ON -DCMAKE_COMMAND="$(brew --prefix cmake)/bin/cmake"
```

**Comment by carlocab** [comment]
I've been bitten by this twice recently, so:

https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9978
#197088

**Comment by bradking** [comment]
This has been fixed in upstream CMake's post-3.31 development by [CMake MR 9987](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9987).
